sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,634 @@
+        builder.addIfNotNull(DisplayData.item("shardUrl", replica.coreUrl()));
+      }
+    }
+
+    @Override
+    public BoundedReader<SolrDocument> createReader(PipelineOptions options)
+        throws IOException {
+      return new BoundedSolrReader(this);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<SolrDocument> getDefaultOutputCoder() {
+      return SolrDocumentCoder.of();
+    }
+  }
+
+  private static class BoundedSolrReader extends BoundedSource.BoundedReader<SolrDocument> {
+
+    private final BoundedSolrSource source;
+
+    private AuthorizedSolrClient solrClient;
+    private SolrDocument current;
+    private String cursorMark;
+    private Iterator<SolrDocument> batchIterator;
+    private boolean done;
+
+    private BoundedSolrReader(BoundedSolrSource source) {
+      this.source = source;
+      this.cursorMark = CursorMarkParams.CURSOR_MARK_START;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      if (source.replica != null) {
+        solrClient = source.spec.getConnectionConfiguration()
+            .createClient(source.replica.coreUrl());
+      } else {
+        solrClient = source.spec.getConnectionConfiguration()
+            .createClient();
+      }
+
+      SolrQuery solrParams = getQueryParams(source);
+      try {
+        QueryResponse response = solrClient.query(solrParams);
+        updateCursorMark(response);
+        return readNextBatchAndReturnFirstDocument(response);
+      } catch (SolrServerException e) {
+        throw new IOException(e);
+      }
+    }
+
+    private SolrQuery getQueryParams(BoundedSolrSource source) {
+      String query = source.spec.getQuery();
+      if (query == null) {
+        query = "*:*";
+      }
+      SolrQuery solrQuery = new SolrQuery(query);
+      solrQuery.set(CursorMarkParams.CURSOR_MARK_PARAM, cursorMark);
+      solrQuery.addSort("id", SolrQuery.ORDER.asc);
+      if (source.replica != null) {
+        solrQuery.setDistrib(false);
+      }
+      return solrQuery;
+    }
+
+    private void updateCursorMark(QueryResponse response) {
+      if (cursorMark.equals(response.getNextCursorMark())) {
+        done = true;
+      }
+      cursorMark = response.getNextCursorMark();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (batchIterator.hasNext()) {
+        current = batchIterator.next();
+        return true;
+      } else {
+        SolrQuery solrQuery = getQueryParams(source);
+        try {
+          QueryResponse response = solrClient.query(solrQuery);
+          updateCursorMark(response);
+          return readNextBatchAndReturnFirstDocument(response);
+        } catch (SolrServerException e) {
+          throw new IOException(e);
+        }
+      }
+    }
+
+    private boolean readNextBatchAndReturnFirstDocument(QueryResponse response) {
+      if (response.getResults().isEmpty() || done) {
```

Comment:
Does an empty response always mean that the query is done, or do we need to keep reading until nextCursorMark equals current cursor mark?